src/plugins/point_cloud/PointCloud.hh

```diff
@@ -0,0 +1,242 @@
+  /// used to color all points with a color gradient according to their values.
+  /// NaN values on the FloatV message aren't displayed.
+  class PointCloud : public ignition::gui::Plugin
+  {
+    Q_OBJECT
+
+    /// \brief List of topics publishing PointCloudPacked messages
+    Q_PROPERTY(
+      QStringList pointCloudTopicList
+      READ PointCloudTopicList
+      WRITE SetPointCloudTopicList
+      NOTIFY PointCloudTopicListChanged
+    )
+
+    /// \brief List of topics publishing FloatV messages
+    Q_PROPERTY(
+      QStringList floatVTopicList
+      READ FloatVTopicList
+      WRITE SetFloatVTopicList
+      NOTIFY FloatVTopicListChanged
+    )
+
+    /// \brief Color for minimum value
+    Q_PROPERTY(
+      QColor minColor
+      READ MinColor
+      WRITE SetMinColor
+      NOTIFY MinColorChanged
+    )
+
+    /// \brief Color for maximum value
+    Q_PROPERTY(
+      QColor maxColor
+      READ MaxColor
+      WRITE SetMaxColor
+      NOTIFY MaxColorChanged
+    )
+
+    /// \brief Minimum value
+    Q_PROPERTY(
+      float minFloatV
+      READ MinFloatV
+      WRITE SetMinFloatV
+      NOTIFY MinFloatVChanged
+    )
+
+    /// \brief Maximum value
+    Q_PROPERTY(
+      float maxFloatV
+      READ MaxFloatV
+      WRITE SetMaxFloatV
+      NOTIFY MaxFloatVChanged
+    )
+
+    /// \brief Point size
+    Q_PROPERTY(
+      float pointSize
+      READ PointSize
+      WRITE SetPointSize
+      NOTIFY PointSizeChanged
+    )
+
+    /// \brief Constructor
+    public: PointCloud();
+
+    /// \brief Destructor
+    public: ~PointCloud() override;
+
+    // Documentation inherited
+    public: void LoadConfig(const tinyxml2::XMLElement *_pluginElem) override;
+
+    /// \brief Callback function for point cloud topic.
+    /// \param[in] _msg Point cloud message
+    public: void OnPointCloud(const ignition::msgs::PointCloudPacked &_msg);
+
+    /// \brief Callback function for point cloud service
+    /// \param[in] _msg Point cloud message
+    /// \param[out] _result True on success.
+    public: void OnPointCloudService(
+        const ignition::msgs::PointCloudPacked &_msg, bool _result);
+
+    /// \brief Get the topic list
+    /// \return List of topics
+    public: Q_INVOKABLE QStringList PointCloudTopicList() const;
+
+    /// \brief Set the topic list from a string
+    /// \param[in] _pointCloudTopicList List of topics.
+    public: Q_INVOKABLE void SetPointCloudTopicList(
+        const QStringList &_pointCloudTopicList);
+
+    /// \brief Notify that topic list has changed
+    signals: void PointCloudTopicListChanged();
+
+    /// \brief Set topic to subscribe to for point cloud.
+    /// \param[in] _topicName Name of selected topic
+    public: Q_INVOKABLE void OnPointCloudTopic(const QString &_topicName);
+
+    /// \brief Callback function for float vector topic.
+    /// \param[in] _msg Float vector message
+    public: void OnFloatV(const ignition::msgs::Float_V &_msg);
```

Comment:
> How much does API matter in these plugins?

We're free to change it as much as we want, because the header isn't installed. All that users get is the `.so` file.

> maybe these can be privatized.

They need to be public for the QML to call them

> seem an ad-hoc data structure to represent this kind of data. 

Agreed in part. There's a reason we chose this format, right? PC fields are indeed more standard, but for the LRAUV use case, this means either:

* a single cloud that holds lots of different data, which may be wasteful if you're only interested in one type of data, 
* or multiple clouds that have XYZ duplicated.

So I think there is a valid use case for this format, even though it does feel unconventional. And I think the plugin can be extended in the future to alternatively support fields, and give users the choice of one or the other in the UI. Similarly, we should extend the plugin in the future to accept doubles, vectors, etc.

Reply:
Yes, the floats are better than point cloud fields, but I was thinking of whether there's a data structure we hadn't thought of in the rush. In particular, I was thinking of organized point clouds with a width, height, and length. Then the ordering of the float array would be entirely up to the user to figure out, with no help from the API. In practice, that means trial and error. It's not the end of the world, it just seems not as well thought-through. It may be a corner case, as point clouds are often not organized. I won't let perfect get in the way of good.